src/Raven.Server/Rachis/Leader.cs

```diff
@@ -1205,41 +1170,42 @@ public class CommandState
         {
             public long CommandIndex;
             public object Result;
-            public ConvertResultAction ConvertResult;
+            public Action<object> WriteResultAction;
             public TaskCompletionSource<(long, object)> TaskCompletionSource;
             public Action<TaskCompletionSource<(long, object)>> OnNotify;
         }
 
-        public class ConvertResultAction
+        private class CtxResultWriter : IDisposable
         {
-            private readonly JsonOperationContext _contextToWriteBlittableResult;
-            private readonly ConvertResultFromLeader _action;
-            private readonly SingleUseFlag _timeout = new SingleUseFlag();
+            private readonly Func<object, object> _writeResultFunc;
+            private readonly SingleUseFlag _invalid = new SingleUseFlag();
 
-            public ConvertResultAction(JsonOperationContext contextToWriteBlittableResult, ConvertResultFromLeader action)
+            public object Result { private set; get; }
+            
+            public CtxResultWriter(Func<object, object> writeResultFunc)
             {
-                _contextToWriteBlittableResult = contextToWriteBlittableResult ?? throw new ArgumentNullException(nameof(contextToWriteBlittableResult));
-                _action = action ?? throw new ArgumentNullException(nameof(action));
+                _writeResultFunc = writeResultFunc;
             }
 
-            public object Apply(object result)
+            public void CopyResult(object result)
             {
                 lock (this)
                 {
-                    if (_timeout.IsRaised())
-                        return null;
-
-                    return _action(_contextToWriteBlittableResult, result);
+                    Result =  _invalid.IsRaised() 
+                        ? null : 
+                        _writeResultFunc.Invoke(result);
                 }
             }
 
-            public void AboutToTimeout()
+            public void InvalidateWriter()
```

Comment:
Is it possible that you invalidate the writer, but still can access the Result here?

Reply:
Yes, that is the idea.
The reader should be the context owner and it should know whether the context is valid or invalid.
Practically the reader's life should be in the scope of the context

Reply:
For next time, maybe a better approach would be to create an `OnDispose` event for the context :-)

Reply:
Sounds really nice :)
But it needed to be done on the `Dispose` of `ReturnRequestContext`.
And it will only change the way we call `InvalidateWriter` 

Next time...